### PR TITLE
Geant4 sensitive detectors update 3

### PR DESCRIPTION
**Pull request by civanch**
This PR is made instead of #21767 Geant4 calorimeter sensitive detectors are updated:
1) use const G4Step* pointer in protected and private interfaces where possible
2) use std::unique_ptr in several places
3) optimized ECalSD run time methods - perform one search over map of logical volumes instead of 3
4) use const pointers and references 
5) removed old commented lines

Results should be unchanged. Clean-up is not yet completed, an extra iteration is needed.

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by kpedro88** [comment]
A large number of differences are reported in the comparisons. Is this just due to code being compiled in a different order?

**Comment by civanch** [comment]
I think there is a bug, no algorithm change is expected.

File: SimG4CMS/Calo/src/ECalSD.cc [hunk 1]
```diff
@@ -286,63 +262,41 @@ uint16_t ECalSD::getDepth(const G4Step * aStep) {
...
-    const G4LogicalVolume* lv = hitPoint->GetTouchable()->GetVolume(0)->GetLogicalVolume();
-    const G4ThreeVector  localPoint = setToLocal(hitPoint->GetPosition(),
-                                                 hitPoint->GetTouchable());
-    double detz     = crystalDepth(lv,localPoint);
-    if (detz<0) detz= 0;
-    return (int)detz/layerSize;
-  }
-  return 0;
+  const double layerSize = 10.; //layer size in mm
+  return (int)crystalDepth/layerSize;
```
Comment (VinInn) [inline SimG4CMS/Calo/src/ECalSD.cc]: please avoid division

File: SimG4CMS/Calo/src/ECalSD.cc [hunk 2]
```diff
@@ -167,95 +167,63 @@ ECalSD::ECalSD(const std::string& name, const DDCompactView & cpv,
...
+  double edep = aStep->GetTotalEnergyDeposit();
+
+  // take into account light collection curve for crystals
+  double weight = 1.;
+  if (suppressHeavy) {
+    TrackInformation * trkInfo = (TrackInformation *)(theTrack->GetUserInformation());
+    if (trkInfo) {
+      int pdg = theTrack->GetDefinition()->GetPDGEncoding();
+      if (!(trkInfo->isPrimary())) { // Only secondary particles
+	double ke = theTrack->GetKineticEnergy()/MeV;
```
Comment (VinInn) [inline SimG4CMS/Calo/src/ECalSD.cc]: can this division be avoided?

**Comment by kpedro88** [comment]
@civanch what is the status?

**Comment by civanch** [comment]
@kpedro88 , I was busy with other things but wish to fix this PR

**Comment by civanch** [comment]
It will be opened in the new branch, closed now.

**Pull request closed without merging**